src/azure-cli-core/azure/cli/core/__init__.py

```diff
@@ -295,15 +350,102 @@ def _get_extension_suppressions(mod_loaders):
                             res.append(sup)
             return res
 
+        def _update_command_index():
+            start_time = timeit.default_timer()
+            INDEX[_COMMAND_INDEX_VERSION] = __version__
+            from collections import defaultdict
+            index = defaultdict(list)
+            for command_name, command in self.command_table.items():
+                # Get the top-level name: <vm> create
+                top_command = command_name.split()[0]
+                # Get module name, like azure.cli.command_modules.vm, azext_webapp
+                module_name = command.loader.__module__
+                if module_name not in index[top_command]:
+                    index[top_command].append(module_name)
+            elapsed_time = timeit.default_timer() - start_time
+            logger.debug("Updated command index in %.3f seconds.", elapsed_time)
+            INDEX[_COMMAND_INDEX] = index
+
+        def _roughly_parse_command(args):
+            # Roughly parse the command part: <az vm create> --name vm1
+            # Similar to knack.invocation.CommandInvoker._rudimentary_get_command, but we don't need to bother with
+            # positional args
+            nouns = []
+            for arg in args:
+                if arg and arg[0] != '-':
+                    nouns.append(arg)
+                else:
+                    break
+            return ' '.join(nouns).lower()
+
+        # Check the command index for (command: [module]) mapping, like
+        # "sql": ["azure.cli.command_modules.sql", "azure.cli.command_modules.sqlvm", "azext_sql"]
+        top_command = None
+        # This list contains both built-in modules and extensions
+        index_modules = []
+        index_builtin_modules = []
+        index_extensions = []
+
+        # Clear the tables to make this method idempotent
+        self.command_group_table = {}
+        self.command_table = {}
+
+        # If the command index version doesn't match the current command version, invalidate the command index
+        index_version = INDEX[_COMMAND_INDEX_VERSION]
+        if not (index_version and index_version == __version__):
+            logger.debug("Command index version is invalid or outdated.")
+            invalidate_command_index()
```

Comment:
Added version information in the command index. 

```json
{
    "version": "2.7.0",
    "commandIndex": {
        "acr": [
            "azure.cli.command_modules.acr"
        ],
```

CLI version may change when
- Users update or downgrade Azure CLI to another version.
- Users install multiple versions and use them alternately.
    - This includes a dev environment configured side-by-side with the released version.

When a command is run with an incompatible version (invalid or mismatch), we manually call `invalidate_command_index` to force an update to avoid all possible edge cases.

@fengzhou-msft to help verify this is the correct way to use `__version__`.